driver-core/src/main/com/mongodb/connection/SimpleCommandMessage.java

```diff
@@ -16,41 +16,76 @@
+import org.bson.BsonElement;
+import org.bson.BsonString;
 import org.bson.FieldNameValidator;
 import org.bson.io.BsonOutput;
 
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.mongodb.ReadPreference.primary;
+
 /**
  * A command message that uses OP_QUERY to send the command.
  *
  * @mongodb.driver.manual ../meta-driver/latest/legacy/mongodb-wire-protocol/#op-query OP_QUERY
  */
 class SimpleCommandMessage extends CommandMessage {
-    private final boolean slaveOk;
+    private final ReadPreference readPreference;
     private final BsonDocument command;
     private final FieldNameValidator validator;
 
-    SimpleCommandMessage(final String collectionName, final BsonDocument command, final boolean slaveOk, final MessageSettings settings) {
-        this(collectionName, command, slaveOk, new NoOpFieldNameValidator(), settings);
+    SimpleCommandMessage(final String collectionName, final BsonDocument command, final ReadPreference readPreference,
+                         final MessageSettings settings) {
+        this(collectionName, command, readPreference, new NoOpFieldNameValidator(), settings);
     }
 
-    SimpleCommandMessage(final String collectionName, final BsonDocument command, final boolean slaveOk,
+    SimpleCommandMessage(final String collectionName, final BsonDocument command, final ReadPreference readPreference,
                          final FieldNameValidator validator, final MessageSettings settings) {
-        super(collectionName, OpCode.OP_QUERY, settings);
-        this.slaveOk = slaveOk;
+        super(collectionName, getOpCode(settings), settings);
+        this.readPreference = readPreference;
         this.command = command;
         this.validator = validator;
     }
 
     @Override
     protected EncodingMetadata encodeMessageBodyWithMetadata(final BsonOutput bsonOutput, final int messageStartPosition) {
-        bsonOutput.writeInt32(slaveOk ? 1 << 2 : 0);
-        bsonOutput.writeCString(getCollectionName());
-        bsonOutput.writeInt32(0);
-        bsonOutput.writeInt32(-1);
+        BsonDocument commandToEncode;
+        List<BsonElement> extraElements = null;
+        if (useNewOpCode()) {
+            bsonOutput.writeInt32(0);  // flag bits
+            bsonOutput.writeByte(0);   // payload type
+
+            extraElements = new ArrayList<BsonElement>();
+            extraElements.add(new BsonElement("$db", new BsonString(new MongoNamespace(getCollectionName()).getDatabaseName())));
+            if (!isDefaultReadPreference()) {
+                extraElements.add(new BsonElement("$readPreference", readPreference.toDocument()));
+            }
+            commandToEncode = command;
+        } else {
+            bsonOutput.writeInt32(readPreference.isSlaveOk() ? 1 << 2 : 0);
+            bsonOutput.writeCString(getCollectionName());
+            bsonOutput.writeInt32(0);
+            bsonOutput.writeInt32(-1);
+            // TODO: only do this when sending to mongos?
```

Comment:
All the tests pass, so it seems that mongod going back to 2.4 doesn't care if we send it or not.  I'm inclined to leave it as is rather than add a server type to MessageSettings.